src/metric/xentropy_metric.hpp

```diff
@@ -0,0 +1,384 @@
+#ifndef LIGHTGBM_METRIC_XENTROPY_METRIC_HPP_
+#define LIGHTGBM_METRIC_XENTROPY_METRIC_HPP_
+
+#include <LightGBM/utils/log.h>
+#include <LightGBM/utils/common.h>
+
+#include <LightGBM/metric.h>
+
+#include <algorithm>
+#include <vector>
+#include <sstream>
+
+/* 
+ * Implements three related metrics:
+ *
+ * (1) standard cross-entropy that can be used for continuous labels in [0, 1]
+ * (2) "intensity-weighted" cross-entropy, also for continuous labels in [0, 1]
+ * (3) Kullback-Leibler divergence, also for continuous labels in [0, 1]
+ *
+ * (3) adds an offset term to (1); the entropy of the label
+ *
+ * See xentropy_objective.hpp for further details.
+ *
+ */
+
+namespace LightGBM {
+
+  // label should be in interval [0, 1];
+  // prob should be in interval (0, 1); prob is clipped if needed
+  inline static double XentLoss(float label, double prob) {
+    const double log_arg_epsilon = 1.0e-12;
+    double a = label;
+    if (prob > log_arg_epsilon) {
+      a *= std::log(prob);
+    } else {
+      a *= std::log(log_arg_epsilon);
+    }
+    double b = 1.0f - label;
+    if (1.0f - prob > log_arg_epsilon) {
+      b *= std::log(1.0f - prob);
+    } else {
+      b *= std::log(log_arg_epsilon);
+    }
+    return - (a + b);
+  }
+
+  // hhat >(=) 0 assumed; and weight > 0 required; but not checked here
+  inline static double XentLoss1(float label, float weight, double hhat) {
+    return XentLoss(label, 1.0f - std::exp(-weight * hhat));
+  }
+
+  // Computes the (negative) entropy for label p; p should be in interval [0, 1];
+  // This is used to presum the KL-divergence offset term (to be _added_ to the cross-entropy loss).
+  // NOTE: x*log(x) = 0 for x=0,1; so only add when in (0, 1); avoid log(0)*0
+  inline static double YentLoss(double p) {
+    double hp = 0.0;
+    if (p > 0) hp += p * std::log(p);
+    double q = 1.0f - p;
+    if (q > 0) hp += q * std::log(q);
+    return hp;
+  }
+
+//
+// CrossEntropyMetric : "xentropy" : (optional) weights are used linearly
+//
+class CrossEntropyMetric : public Metric {

```

Comment:
Binary logloss is the special case of this when all targets are either 0 or 1.
This does "regression" on probabilities in general with continuous targets in [0, 1].
It can be thought of as "infinite-data" binary logloss when you have actual probabilities to give the learning algorithm. Binary logloss tries to count the number of pos. and neg. classes, whereas this never does because it makes no sense if you use the [0, 1] targets. The objective instead becomes a "distance" to a target distribution. Sometimes this function-fitting mode is what you want in my experience.